File: pkg/foodweb/src/Intervality.f
! =============================================================================
!
! CALCULATES FOOD WEB INTERVALITY AS FORMULATED BY 
! STOUFFER et al. 2006 (PNAS 103:19015)
!
! Implemented in R: Yangchen Lin  linyangchen@gmail.com
! Implemented in FORTRAN: Karline Soetaert
! 28 July 2011
!
! =============================================================================
      subroutine rprint(msg)
      character (len=*) msg
            call dblepr(msg, 150, 0, 0)
      end subroutine 


! -----------------------------------------------------------------------------
! Estimate sum of diet gaps for a certain flow matrix.
! -----------------------------------------------------------------------------

       SUBROUTINE DietGap (Flow, nrow, ncol, index, work, sumgap)
       IMPLICIT NONE
       INTEGER, INTENT (IN) :: nrow, ncol
       INTEGER, INTENT (IN) :: index(nrow) ! current permutation of rows of Flow
       DOUBLE PRECISION, INTENT (IN) :: Flow(nrow, ncol)
       DOUBLE PRECISION :: work(nrow)
       INTEGER, INTENT (OUT) :: sumgap
       INTEGER :: I, J, Istart, Istop 
       CHARACTER (LEN = 150) :: MSG
       
       SUMGAP  = 0
       
       DO J = 1, ncol
         DO I = 1, nrow
            work(i) = Flow(index(I), J)
         ENDDO                

         ! first nonzero element of flow
         istart = 0
         DO I = 1, nrow
           IF (work(I) .GT. 0) THEN
                istart = I
                EXIT
           ENDIF
         ENDDO
         
         ! last nonzero element of flow
         istop = 0
         DO I = nrow, 1, -1
           if (work(I) .GT. 0) THEN
                istop = I
                EXIT
           ENDIF
         ENDDO
         ! number of intermediate nonzero elements
         IF (istart .GT. 0 .AND. istop .GT. 0) THEN
           DO I = istart, istop
             IF (work(I) .EQ. 0.d0) THEN
               SUMGAP = SUMGAP + 1
             ENDIF
           ENDDO
         ENDIF   
       
!         write(msg,*) J, istart, istop,SUMGAP
!         CALL rwarn(msg)

       ENDDO ! with I
                    
       END SUBROUTINE DietGap

! -----------------------------------------------------------------------------
! Swapping rows
! -----------------------------------------------------------------------------
            
       SUBROUTINE SwapRows(Nrow, Index, Swap1, Swap2)    
       INTEGER, INTENT (IN) :: Nrow, Swap1, Swap2
       INTEGER, INTENT(INOUT) :: Index(Nrow)
       
       INTEGER :: Itmp
       
         Itmp         = Index(Swap1)
         Index(Swap1) = Index(Swap2)
         Index(Swap2) = Itmp
       
       END SUBROUTINE SwapRows                 

! -----------------------------------------------------------------------------
! INTERVALITY INDEX CALCULATION
! -----------------------------------------------------------------------------
            
       SUBROUTINE intervality(Flow, Nrow, Ncol, Swapind, Nswaps, Nloops,       &
     &  Niter, Verbose, StartTemp, Cool, runif, work, intervalSteps,           &
     &  index, bestindex, workindex, Bestinterval)

       IMPLICIT NONE
       INTEGER :: Nrow, Ncol, Nloops, Nsteps, Nswaps, Niter, Verbose
       INTEGER :: Swapind(Nswaps, 2), intervalSteps(Niter,Nloops)
       INTEGER :: Index(Nrow), Bestindex(Nrow), WorkIndex(Nrow)
       DOUBLE PRECISION :: Flow(Nrow, Ncol), work(Nrow), runif(Nswaps)                 
       DOUBLE PRECISION :: StartTemp, Cool
       
       INTEGER :: I, J, Iswap, swap1, swap2
       INTEGER :: StartInterval, WorkInterval, NewInterval, Bestinterval
       LOGICAL :: finished, accept
       DOUBLE PRECISION :: SumGap, T, Draw, Paccept
       CHARACTER *150 :: MSG

  ! initial interval index
         CALL DietGap (Flow, Nrow, Ncol, index, work, StartInterval)
 	       WorkInterval = StartInterval
 	       BestInterval = StartInterval
         
 	! SIMULATED ANNEALING TO MINIMIZE WorkInterval

	      T = StartTemp

   	    finished = .FALSE.
	      if (StartInterval .EQ. 0) finished = .TRUE.
	
	      Iswap = 1
        DO I = 1, Nloops
          IF (finished) EXIT 

  		    DO J = 1, Niter
			
			! swap two random species (rows)

            swap1 = swapind(iswap, 1)  ! or index(swapind(iswap, 1))
	    		  swap2 = swapind(iswap, 2)
            CALL SwapRows(Nrow, index, swap1, swap2)   ! changes index
            CALL DietGap (Flow, Nrow, Ncol, index, work, NewInterval)
            
 			      accept = .FALSE.
            
            IF (NewInterval .LE. WorkInterval) THEN
			        accept = .TRUE.
  		  	  ELSE ! probability of acceptance as defined by Stouffer et al.
	    			  Paccept = exp(-dble(NewInterval-WorkInterval)/T)
  	  			  draw = runif(Iswap)
			  	    IF (draw .LT. Paccept) THEN
                accept = .TRUE.
              ENDIF
            ENDIF     
			  
            Iswap = Iswap +1 
            if (Iswap > Nswaps) Iswap = 1

            IF (accept) THEN
  	  			  WorkInterval = NewInterval

		    	  	IF (WorkInterval .LT. BestInterval) THEN
			    	    BestInterval = WorkInterval
		  		      BestIndex  = Index
    			    ENDIF	

      			ELSE ! restore index
              CALL SwapRows(Nrow, index, swap2, swap1)
            ENDIF 
            				
   			    intervalSteps(J, I) = WorkInterval
    		    IF (WorkInterval == 0) THEN
              finished = .TRUE.
              EXIT
            ENDIF 			
		      ENDDO ! J
 		  IF (verbose ==1) THEN
         write(msg, *) 'annealing temperature', T,
     &   'current best index', BestInterval
         CALL Rprint (msg)
      ENDIF
		      T = T * Cool
        ENDDO ! I
        

       END SUBROUTINE

File: pkg/foodweb/R/Intervality.R
## =============================================================================
##
## CALCULATES FOOD WEB INTERVALITY AS FORMULATED BY 
## STOUFFER et al. 2006 (PNAS 103:19015)
##
## Implemented: Yangchen Lin
## Department of Zoology, University of Cambridge
## linyangchen@gmail.com
## 28 July 2011
##
## =============================================================================


## -----------------------------------------------------------------------------
## SUBSIDIARY FUNCTIONS
## -----------------------------------------------------------------------------

## circumvents problem with sample() when only one item to choose from
## from R documentation of sample(). R >= 2.11.0 only
resample <- function(x, ...) x[sample.int(length(x), ...)]


dietgap <- function (Flow) {
  ## total length of diet gaps of one species (one column)
    gap <- function(x) {
  	# count number of 0s between the first and last positive element
	  pos <- which(x>0)
  	if (length(pos) > 0) 	{
      x <- x[pos[1]:pos[length(pos)]]
		  return(length(which(x == 0)))
    } else return(0)
  }

  return(sum(apply(Flow, 2, gap)))
}  

## -----------------------------------------------------------------------------
## INTERVALITY
## -----------------------------------------------------------------------------

Interval <- function(Flow, 
                     verbose = FALSE, full = FALSE, 
                     control = list()) {	

# ------------------------
# Check validity of input
# ------------------------

  useDLL <- TRUE ######### TESTING: Set to false to use R-code

  if (useDLL) {
    return(Interval_DLL(Flow, full, verbose, control))
  }
  if(is.null(rownames(Flow)) | is.null(colnames(Flow)))
		stop("Error: 'Flow' has no row names and/or column names")

  # simulated annealing parameters: similar code as in optim()
  con <- list(StartTemp = 0.2, Cool = 0.99, 
              Numiter = 1000, EndTemp = 0.1)   
  nmsC <- names(con)
  con[(namc <- names(control))] <- control
  if (length(noNms <- namc[!namc %in% nmsC])) {
     warning("unknown names in 'control': ", paste(noNms, collapse = ", "))
     warning("See 'details' in Interval help file")
  } 
# ------------------------
# Initialising SANN
# ------------------------

  # initial index
  StartIntInd <- dietgap(Flow)
  
	IntInd  <- StartIntInd
  IntMat  <- Flow
  
# keep 'best' set; sometimes the simmulated annealing forgets the best

	BestInd <- StartIntInd
  BestMat <- Flow

	##SIMULATED ANNEALING TO MINIMIZE IntInd

	IntIndMat <- NULL        
	T <- con$StartTemp
	loopcounter <- 0
	is.finished <- FALSE
	if (StartIntInd == 0) is.finished = TRUE
	
	while(T >= con$EndTemp & !is.finished)	{
		loopcounter <- loopcounter+1
		IntIndMat   <- cbind(IntIndMat, rep(NA, con$Numiter))   

		for(counter in 1:con$Numiter) {
			
			##swap two random species (rows)
			swap <- resample(1:nrow(IntMat), 2)
			newIntMat <- IntMat
			newIntMat[swap,] <- newIntMat[rev(swap),]
			rownames(newIntMat)[swap] <- rownames(newIntMat)[rev(swap)]
		
			## calculate new IntInd
			newIntInd <- dietgap(newIntMat)
		
			accept <- FALSE
      if (newIntInd <= IntInd)
			   accept <- TRUE
			else { ##probability of acceptance as defined by Stouffer et al.
				Paccept <- exp(-(newIntInd-IntInd)/T)
				draw <- runif(1)
				if(draw < Paccept) 
          accept <- TRUE
			}   
      if (accept) {
				IntInd <- newIntInd
				IntMat <- newIntMat
				if (IntInd < BestInd) {
  				BestMat <- IntMat
				  BestInd  <- IntInd
  			}	
			} 				
			IntIndMat[counter, loopcounter] <- IntInd
  		if(IntInd == 0) {
          is.finished <- TRUE
          break
      } 			
		}
		if (verbose) print(paste('annealing temperature', T,
			              'current best index', BestInd))

		T <- T*con$Cool

	}
	if (full)
  	return(list(StartTemp = con$StartTemp, Cool = con$Cool, Numiter = con$Numiter,
              EndTemp = con$EndTemp, SANNIntIndex = IntIndMat,
              StartIntIndex = StartIntInd,
              IntIndex = IntInd, IntMatrix = IntMat))
  else
  	return(list(IntIndex = IntInd, IntMatrix = IntMat))
  
}


##EXAMPLE
